### ISSUE TITLE
Allow Resources to wait for checkAuth resolve before rendering

### ISSUE THREAD

**Issue by luanped** [marmelab/react-admin#5229]
This is a similar issue as the one reported here https://github.com/marmelab/react-admin/issues/4006
The gif image in that issue is still applicable (except no data is actually displayed - just the skeleton of the dashboard)

**What you were expecting:**
When user opens react-admin for the first time, the dashboard should display the login page

**What happened instead:**
When user opens react-admin for the first time, the dashboard is displayed momentarily (without data) until the checkAuth is rejected, which is when user is then redirected to login page

I looked at the demo app provided and noticed that there is a workaround of setTimeout of 500ms to avoid this problem, 
https://github.com/marmelab/react-admin/blob/master/examples/demo/src/dataProvider/rest.ts

But that does not feel correct - is there a way to disable the optimistic rendering for the Admin (i.e. don't display dashboard and then logout after checkAuth fails, but simply just don't display it until it resolves true?)

I'm using latest version 3.8.3 together with react-admin-firebase

Thanks

**Comment by fzaninotto** [comment]
As explained before, this optimistic rendering regarding security is a design decision, favoring returning (and authentified) users over expired users. 

Usually, it's not really a problem for new users, because the implementation of `authProvider.checkAuth()` is synchronous, i.e. it checks the presence of a key in localStorage, which was set previously by `authProvider.login()`, and returns a rejected promise if not found. In these cases, the flash is imperceptible.

I'm wondering why your flash takes longer, and what your `checkAuth` actually does. Can you give more details?

**Comment by luanped** [comment]
Hello,
Thanks for your reply 
Please see below my dummy authProvider implementation 

I totally understand what you explained above, about how the design was to favour speed for returning users over the small flash for new users

However assuming you do not have a valid key in localstorage, and attempt to access to a resource URL such as localhost:3000/users
Using the below authProvider would mean that the dashboard/users would be displayed for 5 seconds, before the checkAuth rejects and redirects user to login page

The 5 seconds is arbitrary chosen to make the issue more noticeable. The actual use case I'm having is when using react-admin-firebase, where the checkAuth is async and even if network is fast there could still be a 0.5s flashing involved (0.5s for the checkAuth to reject from firebase)

Not sure if this is solveable - any help/info will be appreciated

Thanks


```
const authProvider = {
    login: params => new Promise(resolve => {
        setTimeout(() => {
            localStorage.setItem('test', 'test');
            resolve();
        }, 5000);
    }),
    logout: params => {
        localStorage.removeItem('test');
        return Promise.resolve();
    },
    checkAuth: async () => {
        await new Promise((resolve, reject) =>  {
            setTimeout(() => resolve(), 5000);
        });
        return localStorage.getItem('test') ? Promise.resolve() : Promise.reject();
    },
    // checkAuth: params => Promise.resolve(),
    checkError: error => Promise.resolve(),
    getPermissions: params => Promise.resolve(),
};
```

**Comment by fzaninotto** [comment]
Why does the checkAuth have to call the Firebase servers? checkauth should return true if the user was logged (and not logged out) some time before. You can rely on the checkError method to redirect expired logged users to the login page.

**Comment by luanped** [comment]
Hello, I think that is what is implemented here: https://github.com/benwinding/react-admin-firebase/blob/master/src/providers/AuthProvider.ts#L78

Basically the first time it's called it has to wait for the user to return (line 84) - which happens async, which causecs a ~500ms flash

Not sure if this is the right place to ask, but from what you have said so far, do you think the checkAuth implementation there should be different, or is correct ?

**Comment by fzaninotto** [comment]
Yes, it's probably a problem in the react-admin-firebase implementation. The checkAuth should just check for the validity of the JWT, and this can be sync (and immediate).

**Comment by thekevinbrown** [comment]
We have this problem because we need to get a CSRF token from the server. It works as follows:

- Session Storage is used instead of Local Storage (because that's the point of CSRF protection in the first place, blindly posting in a new tab in the user's browser shouldn't have the token.)
- A user logs in, then opens a new browser tab and goes to the admin area.
- At this point they have a valid session cookie, but not a valid CSRF token in the new tab.
- The admin area needs to fetch `/current-user` to get the CSRF token so it can be used in future requests, otherwise if we send anything first we'll get an error and the user is logged out.
- We're currently doing this in `checkAuth`.
- This results in about a 500ms to 1 second lag between the admin area showing the first time with an empty UI before we get "Insufficient permissions".
- If the user is already logged in and opens a new tab, it works largely how we'd expect.

Proposed change is while awaiting for `checkAuth` a loading screen is shown instead of the full UI. It looks really unnatural to users at the moment. Alternatively if there's a better place we could be doing this we'd be happy to move the logic elsewhere so it can be synchronous, just need to know where it should go.

**Comment by thecopy** [comment]
I also have this issue, with a twist causing UI failures for users. We use OIDC, and when access-token has expired we want to do a refresh-token exchange to get a new one. However, since react-admin does not wait, we get failed 401 responses. I have circumvented the logout by also doing a refresh in checkError, but the UX is very bad as the user is presented with errors and he is confused.

@fzaninotto 
> optimistic rendering regarding security is a design decision, favoring returning (and authentified) users over expired users.
> 
> Usually, it's not really a problem for new users, because the implementation of `authProvider.checkAuth()` is synchronous, i.e. it checks the presence of a key in localStorage, which was set previously by `authProvider.login()`, and returns a rejected promise if not found. In these cases, the flash is imperceptible.
> 

If "usually" the "flash is imperceptible.", why is it necessary to call the API optimistically?

**Comment by thekevinbrown** [comment]
@thecopy We've worked around this by doing this in our case in the `HttpClient`:

```typescript
const httpClient = (url: string, options: any = {}) => {
	// If we have a new tab, we might be logged in and have a session,
	// but not have an xsrfToken value in our session storage.
	// In this case, go get it real quick first, otherwise we'll trigger
	// a 403, which is a whole pile of pain, logging the user out and all
	// sorts. React Admin doesn't wait for checkAuth before firing off
	// the actual request for the list data.
	//
	// This call caches in sessionStorage, so as soon as this has been called
	// once future calls are very quick.
	return AuthProvider.checkAuth().then(() => {
		if (!options.headers) {
			options.headers = new Headers({
				Accept: 'application/json',
				'Xsrf-Token': sessionStorage.getItem('xsrfToken') ?? '',
			});
		}
		options.credentials = 'include';
		return fetchUtils.fetchJson(url, options);
	});
};

export const DataProvider = simpleRestProvider(process.env.REACT_APP_BASE_URL, httpClient);
```

It's still not great because as I mentioned the empty UI shows for 4-5 seconds in some cases before they get a login screen, but at least the user doesn't get barraged with errors.

**Comment by fzaninotto** [comment]
The only "simple" solution I see in your case is to use an alternative `<Resource>` component that doesn't use `<WithPermissions>` (which renders routes optimistically) but a component of your own. This is doable in user land but requires copying quite a lot of code. 

I agree that it would make sense to have `<Resource>` elements choose whether they render optimistically or not. Something like a `waitForAuth` prop:

```jsx
<Resource name="books" list={BookList} waitForAuth />
```

This param would be passed to the `WithPermissions` component, which would then choose a sync auth strategy instead of an optimistic one. 

So I'm turning this into a feature request. We don't need it in our projects for now, so the core team won't be working on it in the short term. If you need this change, feel free to open a PR to change the `<Resource>` component accordingly, it shouldn't be too hard!

**Comment by rockallite** [comment]
Here's my attempt to disable the optimistic rendering for the Admin (redirect to `/login` on the first mount if it is not authenticated, instead of rendering the dashboard in a flash):

```jsx
// AuthOnlyAdmin.js
import React, { useEffect, useState } from 'react';
import { useHistory, useLocation } from "react-router-dom";
import { AdminContext, AdminUI, useAuthProvider } from 'react-admin';

const AuthOnlyAdminUI = (props) => {
    const authProvider = useAuthProvider();
    const location = useLocation();
    const history = useHistory();
    const [authChecked, setAuthChecked] = useState(false);

    useEffect(() => {
        authProvider.checkAuth()
            .catch((err) => {
                if (location.pathname !== '/login') {
                    history.push('/login');
                }
                return err;
            })
            .finally(() => {
                setAuthChecked(true);
            });
    }, [authProvider, location, history]);

    return authChecked ? <AdminUI {...props} /> : null;
};

const AuthOnlyAdmin = (props) => {
    const {
        authProvider,
        dataProvider,
        i18nProvider,
        history,
        customReducers,
        customSagas,
        initialState,
        logoutButton,
        children,
        ...restProps
    } = props;
    return (
        <AdminContext
            authProvider={authProvider}
            dataProvider={dataProvider}
            i18nProvider={i18nProvider}
            history={history}
            customReducers={customReducers}
            customSagas={customSagas}
            initialState={initialState}
        >
            <AuthOnlyAdminUI
                {...restProps}
                logout={authProvider ? logoutButton : undefined}
            >
                {children}
            </AuthOnlyAdminUI>
        </AdminContext>
    );
};

export default AuthOnlyAdmin;
```

Then use `<AuthOnlyAdmin>` instead of `<Admin>`:
```jsx
import React from 'react';
import { Resource } from 'react-admin';

import AuthOnlyAdmin from "./AuthOnlyAdmin";
import Dashboard from './Dashboard';
import { UserList, UserIcon } from './UserList';
import { UserEdit } from './UserEdit';
import authProvider from './authProvider';
import dataProvider from './dataProvider';
import i18nProvider from './i18nProvider';
import { appName } from './appInfo';


const App = () => {
    return (
        <AuthOnlyAdmin
            authProvider={authProvider}
            dataProvider={dataProvider}
            i18nProvider={i18nProvider}
            disableTelemetry
            title={appName}
            dashboard={Dashboard}
        >
            <Resource name="users" icon={UserIcon} list={UserList} edit={UserEdit} />
        </AuthOnlyAdmin>
    );
};

export default App;
```

**Comment by DieterVDW** [comment]
Hi, I just wanted to weigh in and point out that this behaviour simply looks really unprofessional.
I thought I did something wrong in my implementation, but I'm kind of surprised this behaviour is "by default" .

For all end-users, loading the login screen is their first interaction with your product. Briefly showing the UI and then immediately loading the login is a bad first impression and sets the tone for the perceived quality of your product.
"You've only got one chance to make a first impression" and it's not a great one.
This may be technically irrelevant, but in the real world UX is king. And this just gives a buggy impression, this is not behaviour of a quality web app.

Although I respect the "by design" argument, that's a very technical, "removed from the end-user" argument.
Just from a credibility and perception of quality perspective, I think it would really benefit React Admin to find some kind of workaround for this.
I don't think any serious company is going to accept this behaviour on a customer-facing product.

**Comment by fzaninotto** [comment]
@DieterVDW Thanks for your feedback. You would have done a different tradeoff than us (favoring first impression rather than speed in repeated use) because you probably use react-admin for B2B apps rather than admins. 

If you want an app that doesn't display anything before login, react-admin gives you the ability to do that. You can either use @rockallite's solution, or [ra-rbac](https://marmelab.com/ra-enterprise/modules/ra-rbac), which switches the auth strategyr from optimistic to pessimistic. 

This is probably not well explained in the documentatoin, so we'll need to clarify that.

**Comment by DieterVDW** [comment]
@fzaninotto : Yes we use this customer-facing, so it's a bit awkward for us. Thanks for pointing me to potential solutions, having a look right now!

I'm not super familiar with the internals of RA, so not sure what is technically possible. But just avoiding this behaviour somehow on the very first load would be a great improvement IMO.

I'm not sure what the typical use-case of React Admin is, but I assume that anyone who uses RA will at some point present their work to some other stakeholder, who won't care about the technical explanation of why something happens.

Just to be clear though: React Admin is one hell of an amazing product and I am very very happy with it!
I think this is the only thing that has caught my attention, for the rest developing on RA has been an amazing experience.
I actually learned React while building on React Admin, and I especially appreciate the great documentation.

**Comment by kdeloach** [comment]
I discovered this issue today and agree with others that this behavior is unexpected.

In particular, I was surprised to find that `Resource` pages are fully rendered and the `DataProvider` executes queries (on list views) when accessing resources directly, even though the user is not authenticated.

It doesn't necessarily break anything, but it did make me wonder if I introduced a bug in my custom auth/data provider implementations. I just couldn't understand why the admin was executing queries despite not being authenticated.

To reproduce the issue:

1. Follow the [React-Admin Tutorial](https://marmelab.com/react-admin/Tutorial.html) up to the "Adding a Login Page" step
2. Navigate to http://localhost:3000/

Expected:

Admin should immediately redirect to the login page: http://localhost:3000/#/login

Actual:

Admin executes the list view query (https://jsonplaceholder.typicode.com/users?_end=10&_order=ASC&_sort=id&_start=0) before redirecting to the login page.

**Comment by etapic** [comment]
This should be mentioned in documentation, just wasted half of day to figure where is the problem. Why just don't do by simple logic if auth provider is set, then check first checkAuth() before flashing admin panel.

**Comment by cardeo-ben** [comment]
I agree with @etapic . It seems strange that we are making API requests when we can already confirm locally whether the user is unauthenticated using `checkAuth`.

**Comment by beinoriusju** [comment]
- `react-admin` **should not care** about implementation details of an authentication providers as long as they return Promises.
- `react-admin` **must not leak** any data prior authentication.
- `react-admin` **should show** loading screen by default.

Good product. But this problem is your `achilles` 100%.

Please, suspend your judgement and do the right thing -- WAIT.

**Comment by fzaninotto** [comment]
Fixed by #7475

**Comment by davext** [comment]
This is not fixed by #7475 btw. It still flashes if you don't have `requireAuth`. When you `requireAuth` it prevents users from signing up because there are no props on custom routes to allow anonymous access for specific custom routes. A fix would be adding allowDeauth bool passed to custom routes!

**Comment by fzaninotto** [comment]
@davext I can't reproduce what you describe. With `requireAuth`, both the `login` page and the custom routes with no layout get rendered anyway, which lets users login and register. 

If you can reproduce the error you describe with the `master` branch, please open a new issue and join the link to the reproduction.
